File: ch08/py/bst.py
"""
Implements a binary search tree in python
"""


class BinaryNode(object):
    def __init__(self, left=None, right=None, value=None, parent=None):
        self.left = left
        self.right = right
        self.value = value
        self.parent = parent

    def __eq__(self, other):
        return self.value == other.value

    def __ne__(self, other):
        return self.value != other.value

    def has_sibling(self):
        """
        Returns true if the node has a sibling
        :return:
        """
        return self.get_sibling() is not None

    def get_sibling(self):
        """
        Returns the sibling of the node if it has a sibling

        :return:
        """

        if self.parent.left == self:
            return self.parent.right
        elif self.parent.right == self:
            return self.parent.left
        else:
            return None

    def is_leaf(self):
        """
        Returns true if no children
        :return:
        """
        return self.left is None and self.right is None


class BinarySearchTree(object):
    def __init__(self):
        """
        Init will instantiate an empty tree without even a root node
        """

        self.root = None
        self.num_nodes = 0
        self.height = 0

    def add_root(self, root):
        """
        Adds a root to the tree

        :param root:
        :return:
        """
        if self.root is not None:
            raise AttributeError("Tree already has a root!")

        self.root = root
        self.num_nodes += 1

    def insert(self, item):
        """
        Inserts a value into the correct place in the tree

        :param item:
        :return:
        """

        if self.root is None:
            self.add_root(BinaryNode(value=item))
            return

        def __insert(new_node, node):
            """
            A recursive helper for insertion

            :param item:
            :param node:
            :return:
            """
            # Base case, item already exists (here our tree just stores distinct items)
            if node == new_node:
                raise ValueError("Cannot insert a duplicate item into the tree")

            # Base case, perform the insertion here!
            elif node.is_leaf() or (node.left is None or node.right is None):

                new_node.parent = node

                if item < node.value and node.left is None:
                    node.left = new_node
                elif item > node.value and node.right is None:
                    node.right = new_node
                else:
                    raise Exception("Attempted to insert value at a non-leaf node.")

            # Need to recurse
            else:

                if item < node.value:
                    __insert(new_node, node.left)
                elif item > node.value:
                    __insert(new_node, node.right)
                else:
                    raise Exception("We should never get to this part!")

        node = self.root
        new_node = BinaryNode(value=item)
        __insert(new_node, node)

        # Update our tree metrics
        self.num_nodes += 1
        self.height = self.get_height(self.root)

    def search(self, item, start_node=None, mode="val"):
        """
        Search to see if the item is within the tree
        :param item:
        :return:
        """

        node = self.root if start_node is None else start_node

        # Easiest base case - we found it
        if node.value == item:

            if mode == "val":
                return node.value
            else:
                return node

        # Other base case - we exhausted the search
        elif node.is_leaf() or node is None:
            return None

        # Now the recursive case
        else:

            if item < node.value:
                return self.search(item, node.left, mode=mode)
            elif item > node.value:
                return self.search(item, node.right, mode=mode)

        return self.search(item, node, mode)

    def get_num_nodes(self):
        return self.num_nodes

    def compute_num_nodes(self):
        """
        Count the number of nodes in the tree

        :return:
        """
        counter = 0
        for _ in self:
            counter += 1

        return counter

    def __str__(self):
        """
        Prints values from min to max
        :return:
        """

        out = ""

        def get_gridsize(height, x_pad=4, y_pad=1):
            """
            Helper function which returns width, height of the print grid

            :param x_space:
            :param y_space:
            :param height:
            :return:
            """
            w = (2 ** self.height) * (x_pad + 1)
            h = (height + 1) * (y_pad + 1)

            return w, h

        def get_print_xy(level, node_idx, w, h, y_pad=1):
            """
            Helper function which returns the co-ordinate to print given a level and the node on that level

            :param level:
            :param node:
            :return:
            """
            x = w // ((2 ** level) + 1) * (1 + node_idx)
            y = lvl * (y_pad + 1)

            return int(x), int(y)

        # Size up our grid
        w, h = get_gridsize(self.height)

        # Figure out where we need to plot numbers
        queue = [self.root]
        lvl = 0
        plt = []

        while len(queue) > 0:

            new_queue = []

            for idx, node in enumerate(queue):
                if node is not None:
                    x, y = get_print_xy(lvl, idx, w, h)
                    plt.append(((x, y), node.value))
                    new_queue.extend([node.left, node.right])

            lvl += 1
            queue = new_queue

        # Now print out the tree
        to_plot = plt.pop(0)
        for j in range(h):
            for i in range(w):

                if to_plot[0][0] == i and to_plot[0][1] == j:
                    out += str(to_plot[1])
                    if len(plt) > 0:
                        to_plot = plt.pop(0)

                else:
                    out += " "

            out += "\n"

        return out

    def __iter__(self):
        """
        Returns an inorder iteration through the tree
        :return:
        """

        def inorder(node):
            """
            Recursive helper function for inorder traversal

            :param node:
            :return:
            """

            if node.left is not None:
                yield from inorder(node.left)

            yield node.value

            if node.right is not None:
                yield from inorder(node.right)

        return inorder(self.root) if self.root is not None else iter([])

    def delete_tree(self):
        """
        De-allocates all the nodes and sets the root to None
        :return:
        """

        def __dealloc(node):
            """
            Deallocates a node

            :param node:
            :return:
            """

            if node is None:
                return

            __dealloc(node.left)
            __dealloc(node.right)

            del node

        __dealloc(self.root)

        # Finally reset everything
        self.root = None
        self.num_nodes = 0
        self.height = 0

    def delete_node(self, item, start_node=None):
        """
        Deletes the node corresponding to the value

        :param item:
        :return:
        """

        sn = self.root if start_node is None else start_node
        node = self.search(item, start_node=sn, mode="node")

        if node is None:
            raise Exception("Trying to delete a value that does not exist within the tree.")

        # Now we have 3 possibilities for deletion
        # 1. Node is a leaf, which means we can just delete it
        # 2. Node has one child, in which case we replace the node's value with the child value and delete the child
        # 3. Node has two children, in which case we replace the node's value with the inorder successor
        # and delete the inorder successor node
        if node.is_leaf():
            if node.parent.left == node:
                node.parent.left = None
            elif node.parent.right == node:
                node.parent.right = None

            self.num_nodes -= 1
            self.height = self.get_height(self.root)

        # Node has 1 child with left child being empty
        elif node.left is None:
            node.value = node.right.value
            node.right = None

            self.num_nodes -= 1
            self.height = self.get_height(self.root)

        # Node has 1 child with right child being empty
        elif node.right is None:
            node.value = node.left.value
            node.left = None

            self.num_nodes -= 1
            self.height = self.get_height(self.root)

        # Node is a full node
        # Inorder successor is the min value of the right tree
        else:
            succ_node = self.get_min(node=node.right, mode="node")
            node.value = succ_node.value
            self.delete_node(succ_node.value, start_node=succ_node)

    def get_min(self, node=None, mode="val"):
        """
        Return minimum value, traverse down the left child until we hit the dead end

        :return:
        """
        cur_node = self.root if node is None else node
        while (cur_node.left is not None):
            cur_node = cur_node.left

        if mode == "val":
            return cur_node.value
        elif mode == "node":
            return cur_node
        else:
            raise ValueError("Mode must be one of { val | node }")

    def get_max(self, node=None, mode="val"):
        """
        Return maximum value, traverse down the right child until we hit the dead end

        :return:
        """
        cur_node = self.root if node is None else node
        while (cur_node.right is not None):
            cur_node = cur_node.right

        if mode == "val":
            return cur_node.value
        elif mode == "node":
            return cur_node
        else:
            raise ValueError("Mode must be one of { val | node }")

    def verify_bst(self):
        """
        Check if our binary tree is a valid binary search tree
        i.e.
        1. left subtree is < node & is valid bst
        2. right subtree is > node & is valid bst

        :return:
        """

        def verify_bst_helper(node):
            """
            Recursive helper for verifying the bst

            :param node:
            :return:
            """

            if node.is_leaf():
                return True

            elif node.right is None:
                return node.left.value < node.value and verify_bst_helper(node.left)

            elif node.left is None:
                return node.right.value > node.value and verify_bst_helper(node.right)

            else:
                return node.left.value < node.value and verify_bst_helper(node.left) \
                       and node.right.value > node.value and verify_bst_helper(node.right)

        return verify_bst_helper(self.root) if self.root is not None else False

    def get_before(self, item, mode="val"):
        """
        Returns next highest value in tree before given value, None if None
        :return:
        """

        # The next highest value should be the max of the left branch of the node
        node = self.search(item, start_node=self.root, mode="node")

        if node is None or node.left is None:
            return None

        return self.get_max(node=node.left, mode=mode)

    def get_after(self, item, mode="val"):
        """
        Returns next lowest value in tree after given value. None if None
        :param item:
        :return:
        """
        # The next lowest value in the tree is the min of the right branch
        node = self.search(item, start_node=self.root, mode="node")

        if node is None or node.right is None:
            return None

        return self.get_min(node=node.right, mode=mode)

    def iter_preorder(self):
        """
        Returns a preorder iteration through the tree
        :return:
        """

        def preorder(node):
            """
            Recursive helper for pre-order

            :param node:
            :return:
            """

            yield node.value

            for child in (node.left, node.right):
                if child is not None:
                    yield from preorder(child)

        return preorder(self.root) if self.root is not None else iter([])

    def iter_postorder(self):
        """
        Returns a postorder iteration through the tree
        :return:
        """

        def postorder(node):
            """
            Recursive helper for post-order

            :param node:
            :return:
            """

            for child in (node.left, node.right):
                if child is not None:
                    yield from postorder(child)

            yield node.value

        return postorder(self.root) if self.root is not None else iter([])

    def iter_bfs(self):
        """
        Return the breadth first traversal of the tree
        :return:
        """
        if self.root is None:
            return iter([])

        queue = [self.root]  # This is not an efficient queue structure but it'll do
        while len(queue) > 0:
            node = queue.pop(0)

            if node is not None:
                yield node.value
                queue.extend((node.left, node.right))

    def get_height(self, node=None):
        """
        Returns the height of tree if node = None, else return the height of the tree from that Node

        :param node:
        :return:
        """

        def __get_height(node):
            if node is None or node.is_leaf():
                return 0
            else:
                return 1 + max(__get_height(node.left), __get_height(node.right))

        if node is None:
            return self.height
        else:
            return __get_height(node)

    def balance(self):
        """
        Balances the tree
        :return:
        """

        # To balance the tree we first perform an inorder traversal which should leave us with a sorted array
        arr = list(self.__iter__())
        self.delete_tree()
        self.from_array(arr)

    def from_array(self, arr):
        """
        Build a balanced binary search tree from an array

        :param arr:
        :return:
        """

        def __from_array(arr, parent, child):
            """
            Recursive helper

            :param arr:
            :param parent:
            :return:
            """

            # Check the length of the array to see if we need to do anything as the base case
            if len(arr) < 1:
                return

            else:
                m = len(arr) // 2
                v = arr[m]

                # Assign the correct reference at the parent node
                new_node = BinaryNode(value=v, parent=parent)
                if child == 'left':
                    parent.left = new_node
                elif child == 'right':
                    parent.right = new_node

                # Now recurse
                __from_array(arr[:m], new_node, 'left')
                __from_array(arr[m + 1:], new_node, 'right')

        if self.root is not None:
            raise ValueError("Cannot build a tree as one already exists")

        if len(arr) < 1:
            raise ValueError("Must take an array of at least 1 element")

        # Sort the array
        s_arr = sorted(arr)

        # Extract the middle value of the array
        m = len(s_arr) // 2
        v = s_arr[m]
        self.root = BinaryNode(value=v)

        # Now recurse
        __from_array(s_arr[:m], self.root, 'left')
        __from_array(s_arr[m + 1:], self.root, 'right')

        # Now set the other parameters
        self.num_nodes = len(arr)
        self.height = self.get_height(self.root)

File: ch08/c/bst.c
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include "bst.h"
#include "circarr_queue.h"

/*
* Start with our binary node functions
*/

binary_node *node_new(int value,
                      binary_node *left, binary_node *right, binary_node *parent)
{
    binary_node *new_node = malloc(sizeof(binary_node));
    if (!new_node)
    {
        printf("Unable to allocate new node, exiting...");
        exit(1);
    }

    new_node->left = left;
    new_node->right = right;
    new_node->parent = parent;
    new_node->value = value;

    return new_node;
}

void node_destroy(binary_node *node)
{
    free(node);
}

bool node_has_sibling(binary_node *node)
{
    if (!node_get_sibling(node))
    {
        return true;
    }

    return false;
}

binary_node *node_get_sibling(binary_node *node)
{
    if (node->parent->left == node)
    {
        return node->parent->right;
    }
    else if (node->parent->right == node)
    {
        return node->parent->left;
    }
    else
    {
        return NULL;
    }
}

bool node_is_leaf(binary_node *node)
{
    return (!node->left && !node->right);
}

bool node_is_equal(binary_node *node, binary_node *other_node)
{
    return (node->value == other_node->value);
}

/*
* Now comes all of our tree functions
*/

bst *bst_new(void)
{
    bst *tree = malloc(sizeof(bst));
    if (!tree)
    {
        printf("Unable to allocate new tree!");
        exit(1);
    }

    tree->root = NULL;
    tree->num_nodes = 0;
    tree->height = 0;

    return tree;
}

void bst_add_root(bst *tree, binary_node *node)
{
    if (tree->root)
    {
        printf("Unable to add new root, one already exists!");
    }
    else
    {
        tree->root = node;
        tree->num_nodes += 1;
    }
}

void bst_insert(bst *tree, int value)
{

    binary_node *new_node = node_new(value, NULL, NULL, NULL);

    // If we have an empty tree, we just add the root!
    if (!tree->root)
    {
        bst_add_root(tree, new_node);
        return;
    }
    // Else we use a recursive helper to add the node in the right place
    else
    {
        __bst_insert_helper(new_node, tree->root);
        tree->num_nodes += 1;
        tree->height = bst_get_height(tree, tree->root);
        return;
    }
}

void __bst_insert_helper(binary_node *new_node, binary_node *cur_node)
{
    // Case 1: Item already exist:
    if (new_node->value == cur_node->value)
    {
        printf("Not inserting new node, value already exists...");
    }

    // Case 2: Perform the insertion here as the current node is a leaf node
    else if (node_is_leaf(cur_node))
    {
        new_node->parent = cur_node;

        if (new_node->value < cur_node->value)
        {
            cur_node->left = new_node;
        }
        else if (new_node->value > cur_node->value)
        {
            cur_node->right = new_node;
        }
        else
        {
            printf("Should not reach this code area! Something went wrong!");
            exit(1);
        }
    }

    // Case 3: Current node has free LEFT child and new node value is less
    else if (!cur_node->left && new_node->value < cur_node->value)
    {
        cur_node->left = new_node;
    }

    // Case 4: Current node has free RIGHT child and new node value is more
    else if (!cur_node->right && new_node->value > cur_node->value)
    {
        cur_node->right = new_node;
    }

    // Case 5: We need to recurse
    else
    {
        if (new_node->value < cur_node->value)
        {
            __bst_insert_helper(new_node, cur_node->left);
        }
        else if (new_node->value > cur_node->value)
        {
            __bst_insert_helper(new_node, cur_node->right);
        }
        else
        {
            printf("Something went badly wrong!");
            exit(1);
        }
    }
}

void bst_from_array(bst *tree, int *arr, int arr_len)
{
    if (tree->root)
    {
        printf("Cannot build a tree as the root already exists");
    }

    if (arr_len == 0)
    {
        printf("Must take an array of at least 1 element");
    }

    // Sort the array
    qsort(arr, arr_len, sizeof(int), intcmp);

    // Extract the pivot in the middle of the array
    int m = arr_len / 2;
    int v = arr[m];
    binary_node *root = node_new(v, NULL, NULL, NULL);
    bst_add_root(tree, root);

    // Now prepare the array slices that we will iterate over
    int left_arr_len = m;
    int *left_arr = malloc(sizeof(int) * left_arr_len);
    int right_arr_len = arr_len - m - 1;
    int *right_arr = malloc(sizeof(int) * (right_arr_len));
    if (!left_arr)
    {
        printf("Unable to allocate left_arr");
        exit(1);
    }

    if (!right_arr)
    {
        printf("Unable to allocate right_arr");
        exit(1);
    }

    for (int i = 0; i < left_arr_len; i++)
    {
        left_arr[i] = arr[i];
    }

    for (int i = 0; i < right_arr_len; i++)
    {
        right_arr[i] = arr[i + (m + 1)];
    }

    // Now we can recurse
    __bst_from_array_helper(left_arr, left_arr_len, root, 0);
    __bst_from_array_helper(right_arr, right_arr_len, root, 1);

    // Now set the final parameters
    tree->num_nodes = arr_len;
    tree->height = bst_get_height(tree, tree->root);
}

void __bst_from_array_helper(int *arr, int arr_len, binary_node *parent, int position)
{
    // Base Case
    if (arr_len == 0)
    {
        return;
    }
    else
    {
        // Extract the pivot in the middle of the array
        int m = arr_len / 2;
        int v = arr[m];

        // Now we can set the node
        binary_node *new_node = node_new(v, NULL, NULL, parent);
        // Child is LEFT of parent:
        if (position == 0)
        {
            parent->left = new_node;
        }
        // Child is RIGHT of parent
        else if (position == 1)
        {
            parent->right = new_node;
        }
        else
        {
            printf("This is the shadowlands, Simba should never come here.");
            exit(1);
        }

        // Now prepare the array slices that we will iterate over
        int left_arr_len = m;
        int *left_arr = malloc(sizeof(int) * left_arr_len);
        int right_arr_len = arr_len - m - 1;
        int *right_arr = malloc(sizeof(int) * (right_arr_len));
        if (!left_arr)
        {
            printf("Unable to allocate left_arr");
            exit(1);
        }

        if (!right_arr)
        {
            printf("Unable to allocate right_arr");
            exit(1);
        }

        for (int i = 0; i < left_arr_len; i++)
        {
            left_arr[i] = arr[i];
        }

        for (int i = 0; i < right_arr_len; i++)
        {
            right_arr[i] = arr[i + (m + 1)];
        }

        // Now we can recurse
        __bst_from_array_helper(left_arr, left_arr_len, new_node, 0);
        __bst_from_array_helper(right_arr, right_arr_len, new_node, 1);
    }
}

// Search
int bst_search_val(bst *tree, int item, binary_node *start_node)
{
    binary_node *node = bst_search_node(tree, item, start_node);
    if (node)
    {
        return bst_search_node(tree, item, start_node)->value;
    }
    else
    {
        return -1;
    }    
}

binary_node *bst_search_node(bst *tree, int item, binary_node *start_node)
{
    if (!start_node)
    {
        start_node = tree->root;
    }

    // Base Case - We have found it:
    if (start_node->value == item)
    {
        return start_node;
    }

    // Base Case - We exhausted the search:
    if (!start_node || node_is_leaf(start_node))
    {
        return NULL;
    }

    // Recursive Case
    if (item < start_node->value)
    {
        return bst_search_node(tree, item, start_node->left);
    }
    else if (item > start_node->value)
    {
        return bst_search_node(tree, item, start_node->right);
    }
    else
    {
        printf("We should never get here!");
        exit(1);
    }
}

int bst_get_min(bst *tree, binary_node *start_node)
{
    binary_node *node = bst_get_min_node(tree, start_node);
    if (node)
    {
        return bst_get_min_node(tree, start_node)->value;
    }
    else
    {
        return -1;
    }
    
}

binary_node *bst_get_min_node(bst *tree, binary_node *start_node)
{
    if (!start_node)
    {
        start_node = tree->root;
    }

    while (start_node->left)
    {
        start_node = start_node->left;
    }

    return start_node;
}

int bst_get_max(bst *tree, binary_node *start_node)
{
    binary_node *node = bst_get_max_node(tree, start_node);
    if (node)
    {
        return bst_get_max_node(tree, start_node)->value;
    }
    else
    {
        return -1;
    }
    
}

binary_node *bst_get_max_node(bst *tree, binary_node *start_node)
{
    if (!start_node)
    {
        start_node = tree->root;
    }

    while (start_node->right)
    {
        start_node = start_node->right;
    }

    return start_node;
}

int bst_get_before(bst *tree, int item)
{
    binary_node *node = bst_get_before_node(tree, item);
    if (node)
    {
        return bst_get_before_node(tree, item)->value;
    }
    else
    {
        return -1;
    }
    
}

binary_node *bst_get_before_node(bst *tree, int item)
{
    binary_node *node = bst_search_node(tree, item, tree->root);

    if (node && node->left)
    {
        return bst_get_max_node(tree, node->left);
    }
    else
    {
        return NULL;
    }
}

int bst_get_after(bst *tree, int item)
{
    binary_node *node = bst_get_after_node(tree, item);
    if (node)
    {
        return bst_get_after_node(tree, item)->value;
    }
    else
    {
        return -1;
    }
    
}

binary_node *bst_get_after_node(bst *tree, int item)
{
    binary_node *node = bst_search_node(tree, item, tree->root);

    if (node && node->right)
    {
        return bst_get_min_node(tree, node->right);
    }
    else
    {
        return NULL;
    }
    
}

// Compute key stats
int bst_get_num_nodes(bst *tree)
{
    // Perform an inorder traversal to get the array
    darray *arr = bst_traverse(tree, 0);
    return arr->n;
}

int bst_get_height(bst *tree, binary_node *start_node)
{
    // Base Case
    if (!start_node || node_is_leaf(start_node))
    {
        return 0;
    }

    // Recursive Case
    else
    {
        int left_height = bst_get_height(tree, start_node->left);
        int right_height = bst_get_height(tree, start_node->right);
        if (left_height < right_height)
        {
            return 1 + right_height;
        }
        else
        {
            return 1 + left_height;
        }
    }
}

// Traversal functions
darray *bst_traverse(bst *tree, int type)
{
    darray *res = darray_new(4);

    // Guard against empty trees
    if (!tree->root)
    {
        return res;
    }

    // Use the type of traversal to switch into the right function
    switch (type)
    {
    // Inorder
    case 0:
        __inorder_rec(tree->root, res);
        break;

    // Preorder
    case 1:
        __preorder_rec(tree->root, res);
        break;

    // Postorder
    case 2:
        __postorder_rec(tree->root, res);
        break;

    // BFS
    case 3:
        __bfs_helper(tree, res);
        break;
    }

    return res;
}

void __inorder_rec(binary_node *node, darray *arr)
{
    if (node->left)
    {
        __inorder_rec(node->left, arr);
    }

    darray_append(arr, node->value);

    if (node->right)
    {
        __inorder_rec(node->right, arr);
    }
}

void __preorder_rec(binary_node *node, darray *arr)
{
    darray_append(arr, node->value);

    if (node->left)
    {
        __preorder_rec(node->left, arr);
    }

    if (node->right)
    {
        __preorder_rec(node->right, arr);
    }
}

void __postorder_rec(binary_node *node, darray *arr)
{
    if (node->left)
    {
        __postorder_rec(node->left, arr);
    }

    if (node->right)
    {
        __postorder_rec(node->right, arr);
    }

    darray_append(arr, node->value);
}

void __bfs_helper(bst *tree, darray *arr)
{
    // This could be a pain to implement
    // We need a queue that stores pointer to nodes
    // This needs to be implemented separately before we can do this
    Queue *queue = queue_new(tree->num_nodes);

    queue_enqueue(queue, tree->root);
    while (!queue_isempty(queue))
    {
        binary_node *node = queue_dequeue(queue);

        if (node)
        {
            darray_append(arr, node->value);
            queue_enqueue(queue, node->left);
            queue_enqueue(queue, node->right);
        }
    }
}

// Deletion
void bst_destroy(bst *tree, bool free_mem)
{
    __bst_destroy_node(tree->root);

    if (free_mem)
    {
        free(tree);
    }
    else
    {
        tree->num_nodes = 0;
        tree->height = 0;
        tree->root = NULL;
    }
    
}

void __bst_destroy_node(binary_node *node)
{
    if (!node)
    {
        return;
    }

    __bst_destroy_node(node->left);
    __bst_destroy_node(node->right);

    free(node);
}

void bst_delete_node(bst *tree, int item, binary_node *start_node)
{
    if (!start_node)
    {
        start_node = tree->root;
    }

    binary_node *to_del = bst_search_node(tree, item, start_node);
    if (!to_del)
    {
        printf("Trying to delete a value that does not exist");
    }

    // 3 possibilities for deletion
    // 1. Node is leaf (just delete)
    if (node_is_leaf(to_del))
    {
        if (to_del->parent->left == to_del)
        {
            to_del->parent->left = NULL;
        }
        else if (to_del->parent->right == to_del)
        {
            to_del->parent->right = NULL;
        }
        else
        {
            printf("Unknown error!");
        }

        tree->num_nodes -= 1;
        tree->height = bst_get_height(tree, tree->root);
        free(to_del);
    }

    // 2. Node has one child (left empty)
    // Replace node value with child value and delete the child
    else if (to_del->left == NULL)
    {
        to_del->value = to_del->right->value;

        free(to_del->right);
        to_del->right = NULL;

        tree->num_nodes -= 1;
        tree->height = bst_get_height(tree, tree->root);
    }

    // 2. Node has one child (right empty)
    // Replace node value with child value and delete the child
    else if (to_del->right == NULL)
    {
        to_del->value = to_del->left->value;

        free(to_del->left);
        to_del->left = NULL;
        
        tree->num_nodes -= 1;
        tree->height = bst_get_height(tree, tree->root);
    }

    // 3. Node is a full node
    // Find the inorder successor and delete that
    // Inorder successor is the min value of the right branch
    else
    {
        binary_node *succ = bst_get_min_node(tree, to_del->right);
        to_del->value = succ->value;
        bst_delete_node(tree, succ->value, succ);
    }
}

// Misc
bool bst_verify(bst *tree, binary_node *start_node)
{
    // Guard against empty trees
    if (!tree->root)
    {
        return false;
    }

    // Guard against empty start_nodes
    if (!start_node)
    {
        start_node = tree->root;
    }

    // Now go through each of the cases
    if (node_is_leaf(start_node))
    {
        return true;
    }
    else if (!start_node->right)
    {
        return start_node->left->value < start_node->value 
                && bst_verify(tree, start_node->left);
    }
    else if (!start_node->left)
    {
        return start_node->right->value > start_node->value
                && bst_verify(tree, start_node->right);
    }
    else
    {
        return start_node->left->value < start_node->value
                && bst_verify(tree, start_node->left)
                && start_node->right->value > start_node->value
                && bst_verify(tree, start_node->right);
    }
}

void bst_balance(bst *tree)
{
    // Inorder traversal which leaves us with a sorted array
    darray *arr = bst_traverse(tree, 0);

    // Now we delete the tree except for the containing struct
    bst_destroy(tree, false);

    // Now we create the new tree
    // Now we grab raw array from our dynamic array and the length too
    bst_from_array(tree, arr->arr, arr->n);
}

void bst_print(bst *tree)
{
    // Get the overall width and height of our canvas
    int w = 0;
    int h = 0;

    w = pow(2, tree->height) * 2 + 1;
    h = (tree->height + 1) * 2 + 1;

    // Plot arrays
    int plt_idx = 0;
    int *plt_x = malloc(sizeof(int) * tree->num_nodes);
    int *plt_y = malloc(sizeof(int) * tree->num_nodes);
    int *plt_val = malloc(sizeof(int) * tree->num_nodes);

    // We do a depth first traversal over the tree to formulate our plot indices
    Queue *queue = queue_new(tree->num_nodes * 2);
    queue_enqueue(queue, tree->root);
    Queue *new_queue = NULL;
    int lvl = 0;
    int idx = 0;
    int x = 0;
    int y = 0;

    while (lvl >= 0)
    {
        // In order to index whitespace properly we should construct a new queue for each level
        new_queue = queue_new(tree->num_nodes * 2);
        idx = 0;

        // Loop through the current queue and append to the plot array
        while (queue->n > 0)
        {
            // Get the next node off the queue
            binary_node *nd = queue_dequeue(queue);

            if (nd)
            {
                // Calculate the x,y co-ordinates for it
                x = w / (pow(2, lvl) + 1) * (1 + idx);
                y = lvl * 2;

                // Append to plot arrays
                printf("Appending: (%i, %i -> %i)\n", x, y, nd->value);
                plt_x[plt_idx] = x;
                plt_y[plt_idx] = y;
                plt_val[plt_idx] = nd->value;
                plt_idx++;
                idx++;

                // Grow out the new queue
                if (nd->left)
                {
                    queue_enqueue(new_queue, nd->left);
                }

                if (nd->right)
                {
                    queue_enqueue(new_queue, nd->right);
                }
                printf("Num in new queue: %i\n", new_queue->n);
            }
        }

        // Now we setup the queue for the next level
        queue_destroy(queue);
        queue = new_queue;
        if (queue->n > 0)
        {
            lvl++;
        }
        else
        {
            lvl = -1;
        }
        
    }

    // Now we read back through our plot indices and plot
    // Reset the index we use to keep track to zero, and go ahead
    plt_idx = 0;
    for (int i = 0; i < h; i++)
    {
        for (int j = 0; j < w; j++)
        {

            if (plt_y[plt_idx] == i && plt_x[plt_idx] == j)
            {
                printf("%i", plt_val[plt_idx]);
                plt_idx++;
            }
            else
            {
                printf(" ");
            }
        }

        printf("\n");
    }
}

/*
* Misc
*/ 

int intcmp (const void * a, const void * b) {
   return (*(int*)a - *(int*)b);
} // Needed for using C's "qsort" function